### PR TITLE
DLA-Future Fortran Interface

### PR DESCRIPTION
**Pull request by RMeli**
Fortran interface for [DLA-Future](https://github.com/eth-cscs/DLA-Future).

The implementation is based on our integration in CP2K (https://github.com/cp2k/cp2k/pull/3143), but it is expanded to complex types and the generalised eigensolver.

The interface is written manually in a single Fortran module. The tests are partially generated using `fypp`.

CI is heavily based on the DLA-Future CI and currently provides CPU tests using either MKL and OpenBLAS/netlib-ScaLAPACK.

This new interface has already been tested with CP2K (see the [RMeli/CP2K/dla-future-fortran](https://github.com/RMeli/cp2k/tree/dla-future-fortran) branch) and will be contributed there once `v0.1.0` is officially released and available in Spack.

### CODE REVIEW

File: spack/packages/intel-oneapi-mkl/package.py
Comment (RMeli) [inline spack/packages/intel-oneapi-mkl/package.py:1]: This file will be removed once https://github.com/spack/spack/pull/43673 is merged.

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Review by mtaillefumier — APPROVED** [review]

File: src/dlaf_fortran.f90 [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+   subroutine dlaf_pssyevd(uplo, n, a, ia, ja, desca, w, z, iz, jz, descz, info)
+      character, intent(in) :: uplo
+      integer, intent(in) :: n, ia, ja, iz, jz
+      integer, dimension(9), intent(in) :: desca, descz
+      integer, target, intent(out) :: info
+      real(kind=sp), dimension(:, :), target, intent(inout) :: a, z
+      real(kind=sp), dimension(:), target, intent(out) :: w
+
+      interface
+         subroutine pssyevd_dlaf(uplo_, n_, a_, ia_, ja_, desca_, w_, z_, iz_, jz_, descz_, info_) &
```
Comment (msimberg) [inline src/dlaf_fortran.f90:217]: Just out of ignorance, is there a particular reason for the naming scheme with `subroutine dlaf_x` and then in the `interface` `subroutine x_dlaf`?

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    add_executable(${TEST} ${TEST}.f90)
+    target_link_libraries(${TEST} PRIVATE 
+        DLAF_Fortran 
+        testutils
+        pxpotrf
+        pxheevd
+        pxhegvx
+    	$<$<BOOL:${DLAF_FORTRAN_WITH_MKL}>:MKL::MKL> 
+    )
+    target_include_directories(${TEST} PUBLIC $<TARGET_PROPERTY:DLAF_Fortran,Fortran_MODULE_DIRECTORY>)
```
Comment (msimberg) [inline test/CMakeLists.txt:34]: Is this a CMake/Fortran limitation that it's not enough to simply do `target_link_libraries(${TEST} PUBLIC DLAF_Fortran)`?

File: test/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+# Distributed Linear Algebra with Future (DLAF)
+#
+# Copyright (c) 2018-2024, ETH Zurich
+# All rights reserved.
+#
+# Please, refer to the LICENSE file in the root directory.
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+set(MPIRUN "" CACHE STRING "MPI run command.")
```
Comment (msimberg) [inline test/CMakeLists.txt:11]: I'd probably namespace this variable, since it's a cache variable:
```suggestion
set(DLAF_FORTRAN_MPIRUN "" CACHE STRING "MPI run command.")
```

**Review by msimberg — CHANGES_REQUESTED** [review]
Only minor comments and questions. Looks very nice to me! I can't really comment on the fortran and test parts though.

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+if(NOT CMAKE_BUILD_TYPE)
+  set(CMAKE_BUILD_TYPE "RelWithDebInfo" CACHE STRING "Build Type" FORCE)
+endif()
+
+if(NOT CMAKE_Fortran_MODULE_DIRECTORY)
+    set(CMAKE_Fortran_MODULE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/fortran_mods)
+endif ()
+
+find_package(DLAF REQUIRED)
```
[the remark this answers is not in the document]
Comment (RMeli) [reply CMakeLists.txt:32]: Good catch, definitely needed.

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+# Copyright (c) 2018-2024, ETH Zurich
+# All rights reserved.
+#
+# Please, refer to the LICENSE file in the root directory.
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+add_library(DLAF_Fortran dlaf_fortran.f90)
+
+target_compile_options(DLAF_Fortran PRIVATE -Wall -Wextra -Wpedantic -Werror)
```
[the remark this answers is not in the document]
Comment (RMeli) [reply src/CMakeLists.txt:13]: Good point. These options have become stricter and stricter during testing/debugging and I agree that `-Werror` should not be imposed to the end-user.

File: src/dlaf_fortran.f90 (hunk 1 above)
Comment (RMeli) [reply src/dlaf_fortran.f90:217]: Not really. It's a personal preference for auto-completion (i.e. if I'm integrating DLAF I prefer to autocomplete on `dlaf` to see what's there), but happy to change it if the consensus is to prefer the other way around.

File: spack/packages/dla-future-fortran/package.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+from spack import *
+
+
+class DlaFutureFortran(CMakePackage):
+    """
+    Fortran interface to the DLA-Future library.
+    """
+
+    homepage = "https://github.com/eth-cscs/DLA-Future-Fortran"
+    url      = "http://www.example.com/MyPackage-1.0.tar.gz"
```
[the remark this answers is not in the document]
Comment (RMeli) [reply spack/packages/dla-future-fortran/package.py:16]: Yes. I was waiting for the first release since I'm not sure what the tag will end up being (i.e. `v0.1.0` or `0.1.0`). I'll probably do the former like in DLA-Future.

File: CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (RMeli) [reply CMakeLists.txt:32]: Mmmh... I think we don't deal properly with DLA-Future versions in CMake. I'll look into it.

[the remark this answers is not in the document]
Comment (RMeli) [reply CMakeLists.txt:32]: I think we need https://github.com/eth-cscs/DLA-Future/pull/1121 in order to do this. If that is the preferred solution, should this be patched for all versions (at least in Spack)?

File: spack/packages/dla-future-fortran/package.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+    depends_on("mpi", when="+test")
+    depends_on("scalapack", when="+test")
+    depends_on("py-fypp", when="+test", type="build")
+
+    def cmake_args(self):
+        args = []
+
+        args.append(self.define_from_variant("BUILD_SHARED_LIBS", "shared"))
+        args.append(self.define_from_variant("DLAF_FORTRAN_BUILD_TESTING", "test"))
```
[the remark this answers is not in the document]
Comment (msimberg) [reply spack/packages/dla-future-fortran/package.py:45]: We have a special `ci-test` variant that enables a bit more than just testing, but that one is not upstream. Otherwise we only have "enable tests on `self.run_tests`": https://github.com/spack/spack/blob/309d3aa1ecf163f219ee0fb5db4d0dd91b28560a/var/spack/repos/builtin/packages/dla-future/package.py#L239.

File: src/dlaf_fortran.f90 (hunk 1 above)
Comment (msimberg) [reply src/dlaf_fortran.f90:217]: No preference for me, I was just curious.

File: spack/packages/dla-future-fortran/package.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (RMeli) [reply spack/packages/dla-future-fortran/package.py:45]: I totally missed this, thanks for pointing it out.

File: src/dlaf_fortran.f90 (hunk 1 above)
Comment (msimberg) [reply src/dlaf_fortran.f90:217]: Right, my question was indeed about why in one case `dlaf` is a prefix and in the other it's a suffix. It makes sense that it's a prefix as a "namespace", in the same way the C-API is prefixed with `dlaf` in DLA-Future. And then I was simply wondering why not calling the other one something like `dlaf_x_interface` or `dlaf_x_` or `dlaf_x_someothersuffix`. But as long as this latter thing (that is at the moment called `x_dlaf`) is an internal detail I don't mind either way.

File: test/CMakeLists.txt (hunk 2 above)
Comment (RMeli) [reply test/CMakeLists.txt]: This line is indeed not necessary.

File: src/dlaf_fortran.f90 (hunk 1 above)
Comment (RMeli) [reply src/dlaf_fortran.f90:217]: Yes, it's an internal detail. I have no good justification for `x_dlaf`. A common pattern I've see is to use `dlaf_*_c`, I can convert to that if it's preferred.

File: spack/packages/dla-future-fortran/package.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (RMeli) [reply spack/packages/dla-future-fortran/package.py]: As discussed offline, I think that the fact that `fypp` is needed only as `build` dependency for test, prevents the use of `self.run_tests` because marking it as `depends_on("py-fypp", type=("build", "test")` would make it a build dependency of `dla-future-fortran` too (which is not the case).

File: spack/packages/intel-oneapi-mkl/package.py
Comment (RMeli) [reply spack/packages/intel-oneapi-mkl/package.py:1]: The PR above has been merged. I'll remove this file and bump the Spack version as soon as there is a new `develop-2023-YY-XX` tag upstream (looks like they are tagged on Sundays).

Comment (RMeli) [reply spack/packages/intel-oneapi-mkl/package.py:1]: Bumped Spack to `develop-2024-04-21`, removed custom `intel-oneapi-mkl` since the changes have been merged upstream.

**Comment by RMeli** [comment]
Bumping the Spack version for the patched `intel-oneapi-mkl` results in CI failures because of https://github.com/spack/spack/pull/41373.

**Review by msimberg — APPROVED** [review]
Tentatively approved. I suppose it makes sense to wait for the DLA-Future 0.4.1 release so that the CMake config can be updated here, but otherwise this looks good to me.

File: .github/check_license.sh [hunk 8]
```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env bash
```
Comment (rasolca) [inline .github/check_license.sh:1]: Don't we need a license on this file as well?

File: CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+#
+# Please, refer to the LICENSE file in the root directory.
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+cmake_minimum_required(VERSION 3.22)
+
+# FIXME: CXX needed by DLA-Future
+# FIXME: C needed by HDF5 (dla-future +hdf5)
+project(DLAF-Fortran VERSION 0.1.0 LANGUAGES Fortran C CXX) # TODO: Remove C and CXX
```
Comment (rasolca) [inline CMakeLists.txt:15]: 🤔 how is it done in other projects?

File: src/dlaf_fortran.f90 (hunk 1 above)
Comment (rasolca) [reply src/dlaf_fortran.f90:217]: Yes, I would go either for `dlaf_*_c` or `dlaf_c_*`

File: spack/packages/dla-future-fortran/package.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+    variant("shared", default=True, description="Build shared libraries.")
+    variant("test", default=False, description="Build tests.")
+
+    generator("ninja")
+    depends_on("cmake@3.22:", type="build")
+
+    # Requires 0.3.0 for complex API
+    # Requires 0.4.0 for Intel OneAPI MKL
+    depends_on("dla-future@0.4.0: +scalapack")
```
Comment (rasolca) [inline spack/packages/dla-future-fortran/package.py:33]: 0.4.1
(as discussed, just an inline reminder)

**Review by rasolca — APPROVED** [review]
Looks good, some minor comments.

File: CMakeLists.txt (hunk 9 above)
Comment (msimberg) [reply CMakeLists.txt:15]: Should DLA-Future be using `enable_language(C CXX)` in its `DLAFConfig.cmake` file perhaps? Or `find_package(HDF5 COMPONENTS C)` or something like that? I'm not sure what the right thing to do is though...

File: .github/check_license.sh (hunk 8 above)
Comment (RMeli) [reply .github/check_license.sh:1]: Sure. But I don't think it is currently tested, and we should add it to the DLA-Future one as well.

File: src/dlaf_fortran.f90 (hunk 1 above)
Comment (RMeli) [reply src/dlaf_fortran.f90]: I changed the naming convention to `dlaf_*_c`.

File: CMakeLists.txt (hunk 9 above)
Comment (RMeli) [reply CMakeLists.txt:15]: I suspect that `FindMPI` does require the `C` and `CXX` languages to be enabled. (`C` only when HDF5 is a dependency of DLA-Future, so we might be able to remove it by making HDF5 private).

File: test/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+# Distributed Linear Algebra with Future (DLAF)
+#
+# Copyright (c) 2018-2024, ETH Zurich
+# All rights reserved.
+#
+# Please, refer to the LICENSE file in the root directory.
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+set(DLAF_FORTRAN_MPIRUN "" CACHE STRING "MPI run command.")
```
Comment (albestro) [inline test/CMakeLists.txt:11]: is there any reason why it is better to have this custom variable instead of using `FindMPI` [variables](https://cmake.org/cmake/help/latest/module/FindMPI.html#variables-for-locating-mpi)?

File: test/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+)
+
+foreach(TEST ${TEST_EXECUTABLES})
+    add_executable(${TEST} ${TEST}.f90)
+    target_link_libraries(${TEST} PRIVATE 
+        DLAF_Fortran 
+        testutils
+        pxpotrf
+        pxheevd
+        pxhegvx
```
Comment (albestro) [inline test/CMakeLists.txt:31]: Just checking: these are just "supports" for the test, right? In principle (it's not a request, just checking my understanding), these should be added as dependency just for the related target, right? (I mean, `p?potrf` tests do not need `pxheevd` and `pxhegvx` as dependency).

File: src/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+add_library(DLAF_Fortran dlaf_fortran.f90)
+
+target_compile_options(DLAF_Fortran PRIVATE -Wall -Wextra -Wpedantic)
+
+target_link_libraries(
+  DLAF_Fortran
+  PRIVATE DLAF::DLAF
```
Comment (albestro) [inline src/CMakeLists.txt:17]: Shouldn't also `MPI::Fortran` be a dependency of `DLAF_Fortran`? If we agree on this, we should probably remove the "explicit" dependency elsewhere.

File: spack/packages/dla-future-fortran/package.py [hunk 14]
```diff
@@ -30,6 +30,7 @@ class DlaFutureFortran(CMakePackage):
 
     # Requires 0.3.0 for complex API
     # Requires 0.4.0 for Intel OneAPI MKL
+    # Requires 0.4.1 for CMake version
```
[the remark this answers is not in the document]
Comment (RMeli) [reply spack/packages/dla-future-fortran/package.py:33]: Thanks, I'll remove this comment. It was meant initially to remind me allow `0.3.0` when MKL is not used, but since we are going to use the fixed CMake version it no longer matters.

File: test/utils/testutils.fpp [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+#:set precision = ['sp', 'dp']
+#:set types = ['real', 'complex']
+#:set symbols = {('sp', 'real'): 's', ('sp', 'complex'): 'c', ('dp', 'real'): 'd', ('dp', 'complex'): 'z'}
+module testutils
+
+#ifdef DLAF_FORTRAN_WITH_MPI_F08
+   use mpi_f08
+#else
+   use mpi
+#endif
```
[the remark this answers is not in the document]
Comment (RMeli) [reply test/utils/testutils.fpp:20]: Ah. Well spotted. Can you please share what the problem/complaint was? I haven't seen it, or I completely missed it.

The F08 version is unfortunately not tested in CI, because the `mpich` version we have to use there is too old. =(

File: test/CMakeLists.txt (hunk 11 above)
Comment (RMeli) [reply test/CMakeLists.txt:11]: It's mostly for convenience, since the `FindMPI` variables are a mouthful.

```
-DDLAF_FORTRAN_MPIRUN="mpiexec -n 6"
```

vs

```
-DMPIEXEC_EXECUTABLE="mpiexec" -DMPIEXEC_NUMPROC_FLAG="-n" -DMPIEXEC_MAX_NUMPROCS="6"
```

I can change it if it's preferred, but if some of those variables are set automatically, it might interfere with CI.

File: test/CMakeLists.txt (hunk 12 above)
Comment (RMeli) [reply test/CMakeLists.txt:31]: Yes, they are not required. I think I was a bit lazy here initially, I'll change it.

File: test/utils/testutils.fpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (albestro) [reply test/utils/testutils.fpp:20]:
```bash
ialberto@daint101:~/workspace/dlaf-fortran/build> make
[  5%] Built target DLAF_Fortran
[  7%] Generating testutils.f90
[ 10%] Building Fortran object test/utils/CMakeFiles/testutils.dir/testutils.f90.o
/users/ialberto/workspace/dlaf-fortran/build/test/utils/testutils.f90:13:2:

   13 | #ifdef DLAF_FORTRAN_WITH_MPI_F08
      |  1
Warning: Illegal preprocessor directive
/users/ialberto/workspace/dlaf-fortran/build/test/utils/testutils.f90:15:2:

   15 | #else
      |  1
Warning: Illegal preprocessor directive
/users/ialberto/workspace/dlaf-fortran/build/test/utils/testutils.f90:17:2:

   17 | #endif
      |  1
Warning: Illegal preprocessor directive
/users/ialberto/workspace/dlaf-fortran/build/test/utils/testutils.f90:14:8:

   14 |    use mpi_f08
      |        1
Fatal Error: Cannot open module file 'mpi_f08.mod' for reading at (1): No such file or directory
compilation terminated.
make[2]: *** [test/utils/CMakeFiles/testutils.dir/build.make:79: test/utils/CMakeFiles/testutils.dir/testutils.f90.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:595: test/utils/CMakeFiles/testutils.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```

File: test/utils/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+foreach(infileName IN LISTS fppFiles)
+
+    string(REGEX REPLACE ".fpp\$" ".f90" outfileName "${infileName}")
+
+    set(outfile "${CMAKE_CURRENT_BINARY_DIR}/${outfileName}")
+    set(infile "${CMAKE_CURRENT_SOURCE_DIR}/${infileName}")
+
+    add_custom_command(
+        OUTPUT "${outfile}"
+        COMMAND fypp "${infile}" "${outfile}"
```
[the remark this answers is not in the document]
Comment (RMeli) [reply test/utils/CMakeLists.txt:22]: Using the `REQUIRED` keyword for `find_program` results in
```
CMake Error at CMakeLists.txt:40 (find_program):
  Could not find FYPP_PROGRAM using the following names: fypp
```
This is not a very informative message IMO and exposes internal details (`FYPP_PROGRAM` variable name). I went for the following custom error message instead:
```
CMake Error at CMakeLists.txt:43 (message):
  fypp (Fortran preprocessor) not found.  fypp is required to build tests.
```

File: src/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+)
+
+add_library(DLAF::DLAF_Fortran ALIAS DLAF_Fortran)
```
[the remark this answers is not in the document]
Comment (RMeli) [reply src/CMakeLists.txt:20]: I think this is needed to add the `DLAF::` namespace, but I'm not 100% sure. I'll double check.

File: test/utils/testutils.fpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (RMeli) [reply test/utils/testutils.fpp]: > When using the [Ninja](https://cmake.org/cmake/help/latest/generator/Ninja.html#generator:Ninja) generator, all source files are first preprocessed in order to generate module dependency information.

The issue is reproducible with `make`.

File: src/CMakeLists.txt (hunk 13 above)
Comment (RMeli) [reply src/CMakeLists.txt:17]: After some discussion, we decided to leave it like this since the interface itself does not depend directly from MPI.

File: src/CMakeLists.txt (hunk 17 above)
[the remark this answers is not in the document]
Comment (RMeli) [reply src/CMakeLists.txt:20]: The `DLAF::` namespace is added by `install`, so this is indeed not needed.

File: test/utils/testutils.fpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (RMeli) [reply test/utils/testutils.fpp]: Since the preprocessor directives are localised in a single file, I enabled the `Fortran_PREPROCESS` property on the associated target only. We can eventually re-visit this if needed. Thank you for spotting this problem!

File: test/CMakeLists.txt (hunk 11 above)
Comment (RMeli) [reply test/CMakeLists.txt]: I changed the CMake logic to allow the use of those variables directly. Ultimately, only `MPIEXEC_MAX_NUMPROCS` needs to be set explicitly. I included a check to ensure that it is set to `6`, which is the number of ranks that the tests require.

I also added a `DLAF_FORTRAN_CSCS_CI` variable (conditional on testing being enabled) and a `+cscs-ci` variant in the Spack package (only for the local recipe) that does the right thing for CI.

File: CMakeLists.txt [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+  if(DLAF_FORTRAN_WITH_MKL)
+    set(MKL_INTERFACE "lp64" CACHE STRING "")
+    set(MKL_THREADING "sequential" CACHE STRING "")
+    set(MKL_MPI "mpich" CACHE STRING "")
+    set(ENABLE_BLACS "ON" CACHE STRING "")
+    set(ENABLE_SCALAPACK "ON" CACHE STRING "")
+    find_package(MKL REQUIRED)
+  else()
+    find_package(ScaLAPACK REQUIRED)
+  endif()
```
[the remark this answers is not in the document]
Comment (RMeli) [reply CMakeLists.txt]: After discussion, I removed the explicit ScaLAPACK dependency from CMake and Spack and use instead `DLAF::SCALAPACK`.

[0c24c6c](https://github.com/eth-cscs/DLA-Future-Fortran/pull/1/commits/0c24c6c692dca68f775788130b509dad3a18cf5b)

File: test/CMakeLists.txt [hunk 19]
```diff
@@ -8,30 +8,56 @@
...
+# Define test executables
+set(INIT_TEST_EXECUTABLES init)
+set(PXPOTRF_TEST_EXECUTABLES pspotrf pdpotrf pcpotrf pzpotrf)
+set(PXHEEVD_TEST_EXECUTABLES pssyevd pdsyevd pcheevd pzheevd)
+set(PXHEGVX_TEST_EXECUTABLES pssygvx pdsygvx pchegvx pzhegvx)
+
+
+find_program(FYPP_PROGRAM fypp)
+
+if("${FYPP_PROGRAM}" MATCHES ".*-NOTFOUND")
```
Comment (albestro) [inline test/CMakeLists.txt:20]: I think it should be exactly the same behavior as (unless I'm missing some edge case)

```suggestion
if(NOT FYPP_PROGRAM)
```

https://cmake.org/cmake/help/latest/command/if.html#basic-expressions

File: test/helpers/CMakeLists.txt [hunk 20]
```diff
@@ -8,18 +8,18 @@
...
-    string(REGEX REPLACE ".fpp\$" ".f90" outfileName "${infileName}")
+    string(REGEX REPLACE ".fypp\$" ".f90" outfileName "${infileName}")
 
     set(outfile "${CMAKE_CURRENT_BINARY_DIR}/${outfileName}")
     set(infile "${CMAKE_CURRENT_SOURCE_DIR}/${infileName}")
 
     add_custom_command(
         OUTPUT "${outfile}"
-        COMMAND fypp "${infile}" "${outfile}"
+        COMMAND "$CACHE{FYPP_PROGRAM}" "${infile}" "${outfile}"
```
Comment (albestro) [inline test/helpers/CMakeLists.txt:22]: Out of curiosity: why are you explicitly targeting the CACHE? In the check I commented above you are using the "generic" access.

If there is no specific reason, I would just go for the "generic" access (which in the end [it might use the cache anyway](https://cmake.org/cmake/help/latest/variable/CACHE.html#cache)) just to not risk strange problems and be sure to use always the same value everywhere.

File: test/CMakeLists.txt [hunk 21]
```diff
@@ -8,30 +8,56 @@
...
+find_package(MPI REQUIRED COMPONENTS Fortran)
 
+if(DLAF_FORTRAN_CSCS_CI)
+  set(MPIEXEC_EXECUTABLE "")
+  set(MPIEXEC_NUMPROC_FLAG "")
+  set(MPIEXEC_MAX_NUMPROCS "")
+  set(MPIEXEC_MAX_PREFLAGS "")
+else()
+  if(NOT (${MPIEXEC_MAX_NUMPROCS} EQUAL 6))
+    message(FATAL_ERROR "Test run with 6 processes, got ${MPIEXEC_MAX_NUMPROCS}. Set -DMPIEXEC_MAX_NUMPROCS=6.")
```
Comment (albestro) [inline test/CMakeLists.txt:33]:
```suggestion
    message(FATAL_ERROR "Tests run with 6 processes, got ${MPIEXEC_MAX_NUMPROCS}. Set -DMPIEXEC_MAX_NUMPROCS=6.")
```

or

```suggestion
    message(FATAL_ERROR "Test runs with 6 processes, got ${MPIEXEC_MAX_NUMPROCS}. Set -DMPIEXEC_MAX_NUMPROCS=6.")
```

File: test/CMakeLists.txt (hunk 19 above)
Comment (RMeli) [reply test/CMakeLists.txt:20]: I tried something similar, but had issues. Mainly because without the `-NOTFOUND` it was going to the `if(<string>` case, and always evaluating to `false`.

Comment (RMeli) [reply test/CMakeLists.txt:20]: I'll try your syntax and see if that works (I might have used something slightly different).

**Review by msimberg — APPROVED** [review]

File: test/CMakeLists.txt (hunk 19 above)
Comment (RMeli) [reply test/CMakeLists.txt:20]: I think I might have erroneously used `if(NOT ${FYPP_PROGRAM})`, thanks for spotting this!

File: test/helpers/CMakeLists.txt (hunk 20 above)
Comment (RMeli) [reply test/helpers/CMakeLists.txt:22]: I think I had issues without this, but likely due to an erroneous placement of `add_subdirectory`. Just re-tested without `CACHE` now, and it works. Fixed.

**Review by albestro — APPROVED** [review]

File: CMakeLists.txt (hunk 9 above)
Comment (RMeli) [reply CMakeLists.txt]: We will revisit this in the future, together with https://github.com/eth-cscs/DLA-Future/issues/1122.

**Review by rasolca — APPROVED** [review]
